src/sentry/static/sentry/app/utils/projects.jsx

```diff
@@ -0,0 +1,296 @@
+    // the server, just fill in with an object with only the slug
+    const projectsOrPlaceholder = Array.from(this.fetchQueue).map(slug =>
+      projectsMap.has(slug) ? projectsMap.get(slug) : {slug}
+    );
+
+    this.setState({
+      fetchedProjects: projectsOrPlaceholder,
+      isIncomplete: this.fetchQueue.size !== projects.length,
+      initiallyLoaded: true,
+      fetching: false,
+      fetchError,
+    });
+
+    this.fetchQueue.clear();
+  };
+
+  /**
+   * If `props.slugs` is not provided, request from API a list of paginated project summaries
+   * that are in `prop.orgId`.
+   *
+   * Provide render prop with results as well as `hasMore` to indicate there are more results.
+   * Downstream consumers should use this to notify users so that they can e.g. narrow down
+   * results using search
+   */
+  loadAllProjects = async () => {
+    const {api, orgId, limit} = this.props;
+
+    this.setState({
+      fetching: true,
+    });
+
+    try {
+      const {results, hasMore} = await fetchProjects(api, orgId, {limit});
+
+      this.setState({
+        fetching: false,
+        fetchedProjects: results,
+        initiallyLoaded: true,
+        hasMore,
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
+
+      this.setState({
+        fetching: false,
+        fetchedProjects: [],
+        initiallyLoaded: true,
+        fetchError: err,
+      });
+    }
+  };
+
+  /**
+   * This is an action provided to consumers for them to update the current projects
+   * result set using a simple search query. You can allow the new results to either
+   * be appended or replace the existing results.
+   *
+   * @param {String} search The search term to use
+   * @param {Object} options Options object
+   * @param {Boolean} options.append Results should be appended to existing list (otherwise, will replace)
+   */
+  handleSearch = async (search, {append} = {}) => {
+    const {api, orgId, limit} = this.props;
+
+    this.setState({fetching: true});
+
+    try {
+      const {results, hasMore} = await fetchProjects(api, orgId, {search, limit});
+
+      this.setState(state => {
+        let fetchedProjects;
+        if (append) {
+          // Remove duplicates
+          fetchedProjects = uniqBy(
+            [...state.fetchedProjects, ...results],
+            ({slug}) => slug
+          );
+        } else {
+          fetchedProjects = results;
+        }
+        return {
+          fetchedProjects,
+          hasMore,
+          fetching: false,
+        };
+      });
+    } catch (err) {
+      console.error(err); // eslint-disable-line no-console
+
+      this.setState({
+        fetching: false,
+        fetchError: err,
+      });
+    }
+  };
+
+  render() {
+    const {slugs, children} = this.props;
+
+    return children({
+      // May not need to expose this?
+      initiallyLoaded: this.state.initiallyLoaded,
```

Comment:
If we don't have a use-case for it right now I would err on exposing less over more in the beginning.

Reply:
Removed.